server/src/main/java/org/elasticsearch/plugins/UberModuleClassLoader.java

```diff
@@ -0,0 +1,281 @@
+                throw new IllegalArgumentException(e);
+            }
+        }
+
+        PrivilegedAction<UberModuleClassLoader> pa = () -> new UberModuleClassLoader(
+            parent,
+            moduleName,
+            jarURLs.toArray(new URL[0]),
+            cf,
+            mparent,
+            moduleDenyList,
+            packageNames
+        );
+        return AccessController.doPrivileged(pa);
+    }
+
+    /**
+     * Constructor
+     */
+    private UberModuleClassLoader(
+        ClassLoader parent,
+        String moduleName,
+        URL[] jarURLs,
+        Configuration cf,
+        ModuleLayer mparent,
+        Set<String> moduleDenyList,
+        Set<String> packageNames
+    ) {
+        super(parent);
+
+        this.internalLoader = new URLClassLoader(jarURLs);
+        // code source is always the first jar on the list
+        this.codeSource = new CodeSource(jarURLs[0], (CodeSigner[]) null);
+        // we need a module layer to bind our module to this classloader
+        this.moduleController = ModuleLayer.defineModules(cf, List.of(mparent), s -> this);
+        this.module = this.moduleController.layer().findModule(moduleName).orElseThrow();
+
+        // Every module reads java.base by default, but we can add all other modules
+        // that are not in the deny list so that plugins can use, for example, java.management
+        mparent.modules()
+            .stream()
+            .filter(Module::isNamed)
+            .filter(m -> "java.base".equals(m.getName()) == false)
+            .filter(m -> moduleDenyList.contains(m.getName()) == false)
+            .forEach(m -> moduleController.addReads(module, m));
+
+        this.packageNames = packageNames;
+    }
+
+    /**
+     * @param moduleName
+     *         The module name; or {@code null} to find the class in the
+     *         {@linkplain #getUnnamedModule() unnamed module} for this
+     *         class loader
+     * @param name
+     *         The <a href="#binary-name">binary name</a> of the class
+     *
+     * @return
+     */
+    @Override
+    protected Class<?> findClass(String moduleName, String name) {
+        if (Objects.isNull(moduleName) || this.module.getName().equals(moduleName) == false) {
+            return null;
+        }
+        return findClass(name);
+    }
+
+    /**
+     * @param name
+     *          The <a href="#binary-name">binary name</a> of the class
+     *
+     * @return
+     */
+    @Override
+    protected Class<?> findClass(String name) {
+        String rn = name.replace('.', '/').concat(".class");
+
+        try (InputStream in = internalLoader.getResourceAsStream(rn)) {
+            if (in == null) {
+                return null;
+            }
+            byte[] bytes = in.readAllBytes();
+            return defineClass(name, bytes, 0, bytes.length, codeSource);
+        } catch (IOException e) {
+            // TODO
+            throw new IllegalStateException(e);
```

Comment:
we should probably log here or something, right?

Reply:
I switched it to `UncheckedIOException`, since that's what we're throwing in `EmbeddedImplClassLoader`.